### DISCUSSION TITLE
Exiting a stack of Modals and tidying them up

### DISCUSSION THREAD

**Discussion by munkii** [PrismLibrary/Prism#3149] [.NET MAUI]
I have the following navigation scenario.

Home -> Settings -> BluetoothDeviceSetup (Modal) -> BluetoothDeviceSetUpConfirmation (Modal)

In BluetoothDeviceSetUpConfirmation the user has two choices

1. Save and Exit
2. Use newly configured device

"Save and Exit" should close down the modals and the user should end up back where they begun the BluetoothDeviceSetup journey (In this example it is Settings but it could be somewhere else.

"Use newly configured device" should take the user to a page that let's then use the newly configured device AND ensure the two modals BluetoothDeviceSetup and BluetoothDeviceSetUpConfirmation are no longer in the stack.

In Xamarin Forms I used to do this by looping through pages and popping them (Wasn't using Prism).

How would I do this in Prism

**Comment by brianlagunas** [comment]
Prism has the concept of relational go-back and navigate, so you can navigate in relation to where you are now.

For example, you could use `NavigateAsync("../../NewPage");`. This would leave you with a stack of `Home -> Settings -> NewPage`

Prism also has absolute navigation where you can replace the entire navigation stack.

For example, you could use `NavigateAsync("/NewPage");`. This would leave you with a stack of `NewPage`.

If you're a Prism Commercial Plus license holder you can use the new GoBackToAsync method.

for example, you could use `GoBackToAsync("Settings");`. Then your stack would look like `Home -> Settings`

Hope this helps

**Reply by munkii** [reply]
> For example, you could use NavigateAsync("../../NewPage");. This would leave you with a stack of Home -> Settings -> NewPage

That looks like exactly what I need, I'll give that a try.

I notice in the PrismFullNavigation sample that when the `PageModalClick` is executed it inserts a new NavigationPage when moving to the Modal page

```
var result = await NavigationService.CreateBuilder()
        .AddNavigationPage(true)
        .AddSegment(nameof(Page1ModalPage), s => s.UseModalNavigation())
        .NavigateAsync();
```

Should I be doing that when I leave `Settings` to start the modal journey with `BluetoothDeviceSetup`?

> for example, you could use GoBackToAsync("Settings");. Then your stack would look like Home -> Settings
I was just looking at that before I read your reply

`await this.NavigationService.GoBackToAsync("Settings", navigationParameters);`

FWIW I have access to that in the current NuGet without "Prism Commercial Plus".

**Reply by munkii** [reply]
@brianlagunas I've come back to this over the last couple of days. 

As stated in this issue [Add support for Relative Navigation '../' within a Modal Stack](https://github.com/PrismLibrary/Prism/issues/2813) we cannot use relative navigation in a Modal.

What I was thinking of doing was walking back over the NavigationStack, getting the name of the first non Modal page and doing 

`GoBackToAsync("NameOfFirstNonModal");`

I saw that with IPageAccessor I can look at the ModalStack and NavigationStack of the current Page. The trouble is inside the ViewModel of the Modal ViewModel the `Navigation.NavigationStack` is empty and the `Navigation.ModalStack` obviusly only gives me the pages in the Modal stack.

Can you think of a way of me getting the name of the first non modal so I can do `GoBackToAsync("NameOfFirstNonModal");`?

**Reply by munkii** [reply]
@brianlagunas I was wondering if you could offer any suggestions for this one. I cannot use relative navigation to such as "../../NewPage" as I am in a ModalStack and as we know you cannot do Relative Navigation within a Modal Stack.

Some more detail of one of the User scenarios:

User wants to capture a reading from a BT device

1. Start capture flow with a MODAL Device Connection view
2. When connected and user ready move to a new MODAL to run the capture
3. Within step 2 there can be up to three captures. Then the user can review those in a new MODAL and choose to save one.

On completion of step 3 we need to get the user back to the View before step 1. The user can initiate step 1 from various Views. Currently Steps 1, 2 and 3 are all loaded on top of each other and there is no need to be able to move back through the flow.

I am not aware of a way of popping Step1 when moving to Step2 and popping 2 when moving to Step3. Is there as we are kinda stuck now.

Thanks
